GUIDE.md

```diff
@@ -0,0 +1,136 @@
+someone less familiar with Ansible (or Jinja2).
+
+## Lazy evaluation of variables
+This is a surprising, sometimes useful, sometimes not [1], feature of
+Jinja2 template expansions in variables in Ansible. Ansible expands
+the templates in the variable value when the variable is used, not
+when it is defined, i.e. variables are evaluated lazily [2]. So one
+can do this:
+```yaml
+vars:
+  foo: "{{ bar }}"
+tasks:
+  - set_fact:
+      bar: quux
+  - name: use foo
+    somemodule:
+      someparameter: "{{ foo }}"
+```
+
+`foo` is defined before the tasks even start, but can refer to a value
+which gets defined only during the execution of tasks! So one can use
+a subsequently defined value in a previous variable declaration and
+things continue to work.
+
+An example of this feature is
+https://github.com/linux-system-roles/storage/pull/81 : when defining
+`unused_disk_subfact`, the variable `unused_disks` is used. But this
+variable is not available yet. It is set by
+`get_unused_disk.yml`. Still, it works, because `unused_disk_subfact`
+is actually expanded only after the call to `get_unused_disk.yml`, and
+`unused_disks` is already defined here.
+
+Another consequence is that one variable defined in `vars:` can refer
+to another variable defined in the same `vars:` block, like here:
+```yaml
+  vars:
+    too_large_size: '{{ (unused_disk_subfact.sectors|int + 1) *
+                        unused_disk_subfact.sectorsize|int }}'
+    unused_disk_subfact: '{{ ansible_devices[unused_disks[0]] }}'
+```
+At the point when `too_large_size` will be used, `unused_disk_subfact`
+will be already known, and the expansion of `too_large_size` will work
+properly.
+
+Another place where this is used is here:
+https://github.com/linux-system-roles/timesync/blob/9050dd95314406236bc8869eac2307a6fa932edc/vars/main.yml#L1
+Role variables are loaded when the role starts and
+`ansible_facts.packages` is not defined yet (it is not among facts
+collected by default, we collect it during the role execution). But
+this is not a problem when defining `__timesync_chrony_version`,
+because the template does not get expanded at this moment - it gets
+expanded only when `__timesync_chrony_version` is used, and if it is
+after we collect the fact, everything is fine.
+
+(Beware though, `set_facts` expands variables, because it is like a
+module call. So in `set_fact` you can not refer to variables whose
+values contain templates referring to yet undefined variables. In this
+way `set_fact` serves as a checkpoint that expands everything that's
+given to it and is not expanded yet. This can be useful, too bad that
+facts are global and can not be unset.)
+
+## Setting variables on `roles`/`import_role` statement
+Variables set on the `roles` or `import_role` statement appear like
+parameters, but actually they are made accessible to the whole
+playbook run, just like role variables set in `vars/` - therefore can
+influence further role invocations.
```

Comment:
I don't know if it's worth notiing that this can be made to happen with include_role too: https://docs.ansible.com/ansible/latest/modules/include_role_module.html#parameter-public

Reply:
I'll try that - does it really have the same effect on the variables given in the `vars` keyword to `include_role` as in the case of `import_role`? The effect can't be exactly the same, because variables given on `import_role` are visible even before this task, while `public` makes the variables visible only after the `include_role` task.

Reply:
Ah, good point.